packages/state-importer/src/main.js

```diff
@@ -0,0 +1,240 @@
+        // format error and throw to CLI error handler
+        throw new StateImportError(
+          'Processing batch failed',
+          error.message || error,
+          this._summary
+        )
+      })
+  }
+
+  _createOrUpdate(
+    newStates: Array<StateData>,
+    existingStates: Array<StateData>
+  ) {
+    return Promise.map(newStates, newState => {
+      const existingState = existingStates.find(
+        state => state.key === newState.key
+      )
+      if (existingState)
+        return this._update(newState, existingState)
+          .then(response => {
+            if (response && response.statusCode === 304)
+              this._summary.unchanged += 1
+            else this._summary.updated += 1
+            return Promise.resolve()
+          })
+          .catch(error => {
+            if (this.continueOnProblems) {
+              this._summary.updateErrorCount += 1
+              this._summary.errors.push(error.message || error)
+              this.logger.error(
+                'Update error occured but ignored. See summary for details'
+              )
+              return Promise.resolve()
+            }
+            this.logger.error(
+              'Process stopped due to error while updating state. See summary for details'
+            )
+            this._summary.updateErrorCount += 1
+            this._summary.errors.push(error.message || error)
+            throw error
+          })
+      return this._create(newState)
+        .then(() => {
+          this._summary.created += 1
+          return Promise.resolve()
+        })
+        .catch(error => {
+          if (this.continueOnProblems) {
+            this._summary.createErrorCount += 1
+            this._summary.errors.push(error.message || error)
+            this.logger.error(
+              'Create error occured but ignored. See summary for details'
+            )
+            return Promise.resolve()
+          }
+          this.logger.error(
+            'Process stopped due to error while creating discount code. See summary for details'
+          )
+          this._summary.createErrorCount += 1
+          this._summary.errors.push(error.message || error)
+          throw error
+        })
+    })
+  }
+
+  _update(newState: StateData, existingState: StateData) {
+    const actions = this.syncStates.buildActions(newState, existingState)
+    // don't call API if there's no update action
+    if (!actions.length) return Promise.resolve({ statusCode: 304 })
+    const service = this._createService()
+    const uri = service.byId(existingState.id).build()
+    const body = {
+      version: existingState.version,
+      actions,
+    }
+    const req = { uri, method: 'POST', body }
+    this.logger.debug('Updating existing code entry')
+    return this.client.execute(req)
+  }
+
+  _create(state: StateData) {
+    const service = this._createService()
+    const uri = service.build()
+    const req = { uri, method: 'POST', body: state }
```

Comment:
Can't this be curried over so we don't have to do this on each create? Feels a bit wasteful? like `createCreator` which returns a function having closed other the before?

Reply:
I had something like that but I refactored it away. I will reimplement it.